[export] Support "custom" metadata field.

zhxchen17 opened the pull request:
Summary:
Add a special field in Graph and Node level metadata called "custom" which should be mapped to a json-serializable object, and we guarantee this field should be always preversed across the following transformations:
1. copy/deepcopy
2. run_decompositions()
3. serialization
4. re-exporting

Test Plan: :test_export -- -r custom_tag

Reviewed By: angelayi

Differential Revision: D60291839

SherlockNoMad commented:
This special field behavior should be documented somewhere...

SherlockNoMad commented on test/export/test_serialize.py:
```diff
@@ -1222,16 +1223,115 @@ def forward(self, x):
+    def test_custom_tag_metadata_decomp(self):
+        class Foo(torch.nn.Module):
+            def forward(self, x):
+                return x + x
```
what is aten.add decomposed into?
Better test with an op that actually decomps.

SherlockNoMad commented on torch/_export/serde/serialize.py:
```diff
@@ -1335,13 +1335,20 @@ def serialize_graph(self, graph_module: torch.fx.GraphModule) -> Graph:
 
+    def serialize_graph_module_metadata(self, meta: Dict[str, str]):
+        ret = {}
+        if custom := meta.get("custom"):
```
what about other meta fields? 

Since meta is a Dict[str, str], all fields can be serialized.

zhxchen17 replied:
I believe meta is not Dict[str, str] and I type it wrong...

SherlockNoMad commented on torch/_export/serde/serialize.py:
```diff
@@ -594,8 +594,8 @@ def export_nn_module_stack(val):
-        if quantization_tag := node.meta.get("quantization_tag"):
-            ret["quantization_tag"] = json.dumps(quantization_tag)
+        if custom := node.meta.get("custom"):
+            ret["custom"] = json.dumps(custom)
```
Is custom object always json serializable? Should make this explicit in the error message

zhxchen17 replied:
yeah will do.

SherlockNoMad reviewed (commented):
lgtm with comment. 

Do you plan to add the register_custom_field() API for nn.Module next?

zhxchen17 commented:
> Do you plan to add the register_custom_field() API for nn.Module next?

@SherlockNoMad no I don't plan to work on it in the near future.

zhxchen17 commented:
@pytorchbot rebase

zhxchen17 committed:
[export] Support "custom" metadata field. (#131912)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/131912

Add a special field in Graph and Node level metadata called "custom" which should be mapped to a json-serializable object, and we guarantee this field should be always preversed across the following transformations:
1. copy/deepcopy
2. run_decompositions()
3. serialization
4. re-exporting

Test Plan:
:test_export -- -r custom_tag

buck2 test 'fbcode//mode/opt' fbcode//bolt/nn/executorch/backends/tests/...

Reviewed By: angelayi

Differential Revision: D60291839
